### DISCUSSION TITLE
P600 Pulse Width Survey

### DISCUSSION THREAD

**Discussion by image-et-son** [image-et-son/p600fw#98] [Q&A] [answered]
In order to correct for typical PW deviations it is important to ge a better understanding of the variations of instruments. Everyone ready to participate and capable of using/trying OS versions alpha 13.* (* marking potential hotfix versions) and handling SysEx patches can contribute by sending in 3 patch syx files prepared as described below.

[P600_PulseWidth_Survey.txt](https://github.com/image-et-son/p600fw/files/7940663/P600_PulseWidth_Survey.txt)

**Comment by matrix12x** [comment]
[pwm.zip](https://github.com/image-et-son/p600fw/files/7945961/pwm.zip)
My three files.

**Reply by image-et-son** [reply]
Thanks, I have your values here (again different):

  | Start | Mid | End
-- | -- | -- | --
A | 192 | 28892 | 61371
B | 0 | 29596 | 61884


But: I found something odd: your 3rd SysEx (patch 22) for the "End" was corrupted. It contained a different number where the "Storage Magic" would be expected. I have no idea how this can happen. Of course, this type of situation would explain failed loads of patches. Does it load properly when selected? Could you please store it under a different number and export it again for me?

**Reply by matrix12x** [reply]
[patch23.syx.zip](https://github.com/image-et-son/p600fw/files/7946192/patch23.syx.zip)

**Reply by matrix12x** [reply]
Of note I set my midpoint, start and end, using an oscilloscope.
I wonder if there is a voice to voice difference also? (I didn't check)
I also wonder if the voltages are different.

If I press 22 it loads fine. I did not try and send it back to the P600.

**Reply by image-et-son** [reply]
> [patch23.syx.zip](https://github.com/image-et-son/p600fw/files/7946192/patch23.syx.zip)

That came out with the correct Storage Magic. Mysterious...

**Comment by el-folie** [comment]
Couldn´t get the python script to work, so here are my alpha13 PW patches:
[PW_start-mid-end_patches.zip](https://github.com/image-et-son/p600fw/files/7946332/PW_start-mid-end_patches.zip)

**Reply by image-et-son** [reply]
| Start | Mid | End
-- | -- | -- | --
A | 192 | 30621 | 61884
B | 256 | 30685 | 62396


BTW: when I made the python script work for Windows I found out the command is "py syx_converter.py <patchfilename>". So just "py", not "python", maybe that's what didn't work in your case.

**Reply by el-folie** [reply]
Yes, the "py" now gets the syx_converter to act, but it spits out another error message:
![p1](https://user-images.githubusercontent.com/27816393/151267451-4e3b0d57-73cb-4d20-b50d-5d39476ed5e5.jpg)

**Reply by matrix12x** [reply]
BTW, the script works fine on my Mac. the command line for me is python3 syx_converter.py patch22.syx  

Note the patch name has to be  ".syx," not ".py"

**Reply by image-et-son** [reply]
Yes, the ending should be syx but the "*.py" ending should also work (the file can have any ending). el-folie's error message is daunting as it points to elementary encoding setting, possibly related to the operating system. I'll look into it when I have time - it would be nice if the script worked for everyone.

**Comment by image-et-son** [comment]
My P600 (remeasured):

  | Start | Mid | End
-- | -- | -- | --
A | 192 | 30144| 60608
B | 192 | 29888| 60416


Unfortunately, no one else contributed so far. For a survey that is a little thin.

**Comment by el-folie** [comment]
Maybe it´s just that a perfect knob mid point/50% duty cycle is not that important for most users. But of course, for pure patch compatibility matters it´d still be nice if more users would send in their data.

**Comment by matrix12x** [comment]
Did you notice your start, mid, and end drift at all as the P600 warmed up?
also, do you want me to recheck mine on alpha 13.2?

Edit:
I remeasured on Alpha 13.2 after leaving on for 30 min. And confirmed PW on the O-scope.
VCOA Start 64
VCOB Start 64

VCOA Mid 30912
VCOB Mid 32192
(mid was tougher to read on the scope, so there is some tolerance here)

VCOA End 62400
VCOB End 61888

**Comment by image-et-son** [comment]
My preliminary conclusion would be

INSTRUMENT | OSC | START | MID | END | MID DEV | RANGE
-- | -- | -- | -- | -- | -- | --
MX | A | 64 | 30912 | 62400 | 1% | 95%
MX | B | 64 | 32912 | 61888 | -6% | 94%
IM | A | 192 | 30144 | 60608 | 0% | 92%
IM | B | 192 | 29888 | 60416 | 1% | 92%
EL | A | 192 | 30621 | 61884 | 1% | 94%
EL | B | 256 | 30685 | 62396 | 1% | 95%


The PW range is limited (first impression: <=95%). It always starts pretty much at zero. The 
MID point is pretty much at half range (@matrix12x : your OSC B mid point seems to be an outlier). 

We could scale the PW with carefully chosen values with the aim to cover the PW range better with the pots. it would b better to have more data, though. What it does not explain is why @el-folie reports that the LFO range is not sufficient to go from zero to zero. The limited voltage range  of PW would make that even easier? 

BTW: in 13.2 the highest LFO amount has slightly increased due to the compatibility fix.

**Reply by matrix12x** [reply]
Re OSC B midpoint, I agree and will retest. it was a little drifty when I was looking on the scope. How are you guys determining midpoint? (eyes or ears?)

**Reply by el-folie** [reply]
Ears. But I also controlled on a VST scope and it was correct 50 %.

Attached this answer as I didn´t get the thread flow at first. ;-) :
Nice chart!

Re LFO PW depth on 13.2, just tried it and yes, it´s possible to reach zero on either side of the waveform. The sound in my head just has a larger timing gap before the waveform starts again. As if the LFO depth was too much and reaching beyond zero and 100% where there is nothing sounding anymore. And if the depth would be deep enough one could even tune the gap timing to the LFO speed. But generally yes, LFO reaches zero to zero now, confirmed!

**Reply by matrix12x** [reply]
From the data sheet (just for capturing in the conversation):
![image](https://user-images.githubusercontent.com/4423119/151665184-d197c09a-68cb-404d-8b57-1d3b2fef44c4.png)

When I retry my mid PW I will check the voltages for start and end on mine.

**Reply by matrix12x** [reply]
Our DAC circuit outputs 0 to +5V. So there may be cases where people cannot hit 0% or 100%. For example if the CEM chip they have needs 5.4V to hit 100% PW. or if it needs -0.015V for 0% PW.

**Reply by el-folie** [reply]
So it´s really a hardware thing and would need trimmers to make each P600 behave perfectly. The DAC gain will also have influence then. Maybe we should just accept the little differences then?

**Reply by matrix12x** [reply]
ok remeasure, I found there is a function on my scope that displays the PW %, and getting 50.0% is a nightmare, lol (stupid scratchy pots). I can get 49.9% or 50.1% reliably though.

PWA mid: 30592
PWB mid: 30720

Also confirmed Alpha 13.2 LFO Amount to PW gets me from zero to zero.

**Reply by el-folie** [reply]
Well, 0.1 % deviation should still be ok. ;-)))

**Reply by image-et-son** [reply]
> From the data sheet (just for capturing in the conversation): ![image](https://user-images.githubusercontent.com/4423119/151665184-d197c09a-68cb-404d-8b57-1d3b2fef44c4.png)
> 
> When I retry my mid PW I will check the voltages for start and end on mine.

What are the three columns in the table? Variations in spec or different chip variants? 

Ah, I see - looked it up myself. Sorry to bother you. The ranges seem too large for an instrument. I guess there must have been cherry picking to use the ones with lower 100% voltage. Our 3 instrument are below 5V, obviously.

**Reply by matrix12x** [reply]
min   typical   max. 

Also, I was looking at the "buffer" after the DAC and before the PW input and it is interesting. Maybe I was wrong about the available voltage range for PW.

![image](https://user-images.githubusercontent.com/4423119/151667325-7d44d51d-d53d-41f6-9194-373937b6dd13.png)

**Reply by matrix12x** [reply]
From the Service Manual:
![image](https://user-images.githubusercontent.com/4423119/151671848-0827c19b-89d1-4e98-a3dc-7d1b075feaa2.png)

**Reply by el-folie** [reply]
Nice! So the circuit makes sure full PW cutoff even at lower DAC gain settings than 5v. As those are 1% resistors, the PW difference between P600s could more likely be CEM3340 batch variation, like Matrix12x already assumed.

**Comment by image-et-son** [comment]
In principle you could replace 410 and 413 with pots to calibrate the PW voltage characteristic. But a things stand I feel that we don't have enough data to make a one-size-fits-all adjustment on the software side, especially when you look at the range in spec. I think we need to decide if the case is strong enough for a calibration procedure (like the one for bender).

I find the explanation for the noise filter interesting. The LFO in the Z80 firmware must be really steppy!?

**Reply by matrix12x** [reply]
You see how low the cutoff frequency is for C443 (0.002uF) and R415 (4.7Mohm)? lol
The Z80 envelopes and LFO were a train wreck.

**Reply by el-folie** [reply]
9.4ms/16.9Hz. Funny, it never appeared to sound bad on Z80. But then again I never listened to the P600 as concentrated as I do since imogen´s great updates. ;-)

**Reply by el-folie** [reply]
>I think we need to decide if the case is strong enough for a calibration procedure (like the one for bender)<

The original idea was just to have the knob´s mid point match the 50% PW - but in the end it´s not that important as the sound engine works as it is. Both zero point cutoffs are reachable. If someone´s bothered too much by the notch to the left on the knob he/I can fiddle with the electronics a bit. 

If a PW cal procedure means again much work for you in coding I don´t know if it´s worth it. It´d be an example of striving for perfection, yes, but not really necessary in this case. The slow detune on fine-tune on the other hand was extremely worth it, can´t get over how good it sounds now, can´t thank you enough. :)

**Reply by matrix12x** [reply]
I don't think we need a PW cal procedure.
Also, yeah, the slow detune is awesome.

**Comment by image-et-son** [comment]
The recent level of bug reporting (and fixing) shows that it is important at some point to converge to a "feature complete" version in which you only fix things but not introduce new issues. Therefore, yes, I feel maybe it would be better to leave it as it is just to make sure we come to an end.

On the other hand, I think it more about patch portability than UI perfection because a PW value obviously leads to different wave shapes on different machines, especially if you create patches which oscillate close to or traverse the far zero point.

**Reply by el-folie** [reply]
That´s true. In that regard it´s a pity that other users didn´t send in their PW syx data, that would have led at least to a better 50% point/shift for the knob.

**Comment by image-et-son** [comment, marked as answer]
OK, I created an enhancement issue for the calibration procedure but marked it "backlog". So it is documented but can wait for another time.

Great excursion here - I am learning so much. I'll close this discussion by marking it "answered".

BTW: I would have marked your posts as "answer" but they were replies, so that was not possible.
